dev/src/order.ts

```diff
@@ -248,19 +248,38 @@ function compareVectors(left: ApiMapValue, right: ApiMapValue): number {
   return compareArrays(leftArray, rightArray);
 }
 
-function stringToUtf8Bytes(str: string): Uint8Array {
-  return new TextEncoder().encode(str);
-}
-
 /*!
  * Compare strings in UTF-8 encoded byte order
  * @private
  * @internal
  */
 export function compareUtf8Strings(left: string, right: string): number {
-  const leftBytes = stringToUtf8Bytes(left);
-  const rightBytes = stringToUtf8Bytes(right);
-  return compareBlobs(Buffer.from(leftBytes), Buffer.from(rightBytes));
+  let i = 0;
+  while (i < left.length && i < right.length) {
+    const leftCodePoint = left.codePointAt(i)!;
+    const rightCodePoint = right.codePointAt(i)!;
+
+    if (leftCodePoint !== rightCodePoint) {
+      if (leftCodePoint < 128 && rightCodePoint < 128) {
+        // ASCII comparison
+        return primitiveComparator(leftCodePoint, rightCodePoint);
+      } else {
+        // Lazy instantiate TextEncoder
+        const encoder = new TextEncoder();
+
+        // UTF-8 encoded byte comparison, substring 2 indexes to cover surrogate pairs
+        const leftBytes = encoder.encode(left.substring(i, i + 2));
```

Comment:
What happens if the character at `i` is a single utf16 code unit but the character at `i+1` is a leading surrogate? In that case, will the encoding still occur correctly?

Reply:
we are going index by index, unless it is surrogate pairs. 
if characters at `i` equals, loop goes to `i+1`, where we start comparison at leading surrogate. 
if they are not equal, we would encode characters at `i` and `i+1`, which falsely included the leading surrogate. but, the characters at index `i` already differs, we would have the result even before the byte code for `i+1` being compared.

Reply:
encoding all remaining substrings are a safe way to go. But also feels like a waste since we are already at the index where it differs. 

We could go for safe solution, but like u mentioned in offline discussion, if the user is storing non-english language into the fields, they would still encounter the performance issue. Because we are basically encoding all strings for them. ie:
```
{text: "这是一个非常长的文本“}
{text: "那是一个非常长的文本“}
... // and other fields that we could have told they are different at the very beginning
```

Reply:
How about this:
```
const leftBytes = encoder.encode(getUtf8SafeSubstring(left, i));

function getUtf8SafeSubstring(str: string, index: number): string {
  const firstCodePoint = str.codePointAt(index)!;
  if (firstCodePoint > 0xffff) {
    // It's a surrogate pair, return the whole pair
    return str.substring(index, index + 2);
  } else {
    // It's a single code point, return it
    return str.substring(index, index + 1);
  }
}
```

In this way, we do not have the magic number 2, and it solved the broken surrogate pair issue as well.

Reply:
I'm not suggesting that there is a bug, and my brain can't reason fully about the logic. And without unit tests it's really hard to see if there is a bug or not. If you port the string generator from https://github.com/firebase/firebase-android-sdk/pull/6706 and the tests pass then I'll feel better :)